btx/processing/freqdir.py

```diff
@@ -0,0 +1,740 @@
+                )
+            for row in X.T:
+                canRankAdapt = numIncorp > (self.ell + 15)
+                if self.nextZeroRow >= self.m:
+                    if self.increaseEll and canRankAdapt and self.rankAdapt:
+                        self.ell = self.ell + 10
+                        self.m = 2*self.ell
+                        self.sketch = np.vstack((*self.sketch, np.zeros((20, self.d))))
+                        self.increaseEll = False
+                        print("INCREASING RANK OF PROCESS {} TO {}".format(self.rank, self.ell))
+                    else:
+                        copyBatch = self.sketch[self.ell:,:].copy()
+                        self.john_rotate()
+                        if canRankAdapt and self.rankAdapt:
+                            reconError = np.sqrt(self.lowMemoryReconstructionErrorUnscaled(copyBatch))
+                            if (reconError > 0.08):
+                                self.increaseEll = True
+                self.sketch[self.nextZeroRow,:] = row 
+                self.nextZeroRow += 1
+                self.num_incorporated_images += 1
+                numIncorp -= 1
+    
+    def john_rotate(self):
+        """ 
+        Apply Frequent Directions Algorithm to 
+        current matrix sketch and adjoined buffer
+
+        Notes
+        -----
+        Based on [1] and [2]. 
+
+        [1] Frequent Directions: Simple and Deterministic Matrix 
+        Sketching Mina Ghashami, Edo Liberty, Jeff M. Phillips, and 
+        David P. Woodruff SIAM Journal on Computing 2016 45:5, 1762-1792
+
+        [2] Ghashami, M., Desai, A., Phillips, J.M. (2014). Improved 
+        Practical Matrix Sketching with Guarantees. In: Schulz, A.S., 
+        Wagner, D. (eds) Algorithms - ESA 2014. ESA 2014. Lecture Notes 
+        in Computer Science, vol 8737. Springer, Berlin, Heidelberg. 
+        https://doi.org/10.1007/978-3-662-44777-2_39
+        """
+        try:
+            [_,s,Vt] = svd(self.sketch , full_matrices=False)
+        except LinAlgError as err:
+            [_,s,Vt] = scipy_svd(self.sketch, full_matrices = False)
+        if len(s) >= self.ell:
+            sCopy = s.copy()
+           #JOHN: I think actually this should be ell+1 and ell. We lose a component otherwise.
+            toShrink = s[:self.ell]**2 - s[self.ell-1]**2
+            #John: Explicitly set this value to be 0, since sometimes it is negative
+            # or even turns to NaN due to roundoff error
+            toShrink[-1] = 0
+            toShrink = sqrt(toShrink)
+            
+            toShrink[:int(self.ell*(1-self.alpha))] = sCopy[:int(self.ell*(1-self.alpha))]
+
+            self.sketch[:self.ell:,:] = dot(diag(toShrink), Vt[:self.ell,:])
+            self.sketch[self.ell:,:] = 0
+            self.nextZeroRow = self.ell
+        else:
+            self.sketch[:len(s),:] = dot(diag(s), Vt[:len(s),:])
+            self.sketch[len(s):,:] = 0
+            self.nextZeroRow = len(s)
```

Comment:
Here too, it'd be super helpful if some explanation were given in the docstring.